### ISSUE TITLE
Broken order of operations when using less variables in calculation and not using less to evaluate it

### ISSUE THREAD

**Issue by battlesnake** [less/less.js#2602]
Demo:

	@a: 1em - 1px;
	@b: 1em + 1px;
	
	sum {
	    actual: calc(@a + 2 * @b);
	    expect: calc(1em - 1px + 2 * (1em + 1px));
	    fix: calc((@a) + 2 * (@b));
	}
	
	diff {
	    actual: calc(@a - 2 * @b);
	    expect: calc(1em - 1px - 2 * (1em + 1px));
	    fix: calc((@a) - 2 * (@b));
	}

Test:

	$ lessc -su=on -sm=on test.less
	sum {
	  actual: calc(1em - 1px + 2 * 1em + 1px);
	  expect: calc(1em - 1px + 2 * (1em + 1px));
	  fix: calc((1em - 1px) + 2 * (1em + 1px));
	}
	diff {
	  actual: calc(1em - 1px - 2 * 1em + 1px);
	  expect: calc(1em - 1px - 2 * (1em + 1px));
	  fix: calc((1em - 1px) - 2 * (1em + 1px));
	}

I understand that this is defined behaviour and not a bug, however it seems a little unintuitive at first so should probably be in the main docs.

**Comment by rjgotten** [comment]
Expressions that can not be further reduced should probably have their output CSS value changed to add enclosing braces. Solves the problem completely, I think.

**Comment by seven-phases-max** [comment]
Yes, it would make sense to document but we don't have such "all that thousands common pitfalls" section there.

Speaking of the code itself: with `-sm=on` it would be almost must-have recommendation to always put parens around complex *variable* values that are supposed to represent an arithmetic expression, e.g::
```less
// -sm:on
@a: 1 + 2; // this has nothing to do with arithmetics at all!
@b: (1 + 2); // now this is an arithmetic expression
```

Also in regards of `sm:on`: see #1872 (and its #1880 follow up). So to be honest I'd recommend to you revert [this](https://github.com/battlesnake/battlemake/commit/0fa82e187c433fde2976cc0099e7fa8168bca2dd) ;) (well, `su:on` does not probably рфму any critical problems, but considering that almost nobody uses this mode you just make you code less-compatible, beside since it not quite used it's also much less tested en masse for possible inconsistencies).

---
Assuming it's not really a compiler issue but more like a documentation suggestion I wonder if this should be moved to `less.js/less-docs` tracker.

**Comment by seven-phases-max** [comment]
@rjgotten
>changed to add enclosing braces

The *only* purpose of `sm:on` is "to not treat values like `font: ... 1em/2px ...;` as arithmetic ops". So by automatically adding parens you'll break those.

**Comment by seven-phases-max** [comment]
OK, before comments start to pile up I'm closing this in favour of #1880 (please do not hesitate to put any ideas there). And for docs suggestion please use [`less.js/less-docs`](https://github.com/less/less-docs).

**Comment by battlesnake** [comment]
`sm:on` also allows one to use CSS `calc`.  I use `su:on` as my background is physics and engineering, and I like dimensional consistency.

**Comment by seven-phases-max** [comment]
Merged to #1880.

**Comment by rjgotten** [comment]
@seven-phases-max 
Oh right. I missed the `strictMath` option there, in its shorthand form. D'oh! Sorry...
